### PR TITLE
Add expect type tests

### PR DESCRIPTION
**Pull request by Thisen**
This PR adds tests of our types. 
This requires that we start running `tsc` in CI.

### CODE REVIEW

File: src/core/atom.ts [hunk 1]
```diff
@@ -17,7 +17,7 @@ type Write<Update> = (
   update: Update
 ) => void | Promise<void>
 
-type WithInitialValue<Value> = {
+export type WithInitialValue<Value> = {
```
Comment (dai-shi) [inline src/core/atom.ts:20]: not exporting this is intentional in #488 as being implementation details. Can we test it without exporting it? I think you can just define it in the test code.

Comment (dai-shi) [reply src/core/atom.ts:20]: Or, as it's not a public type, we don't need to test it?

File: tests/types.test.tsx [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import { expectType } from 'ts-expect'
+import { atom, Atom, WritableAtom } from '../src/index'
+import { WithInitialValue } from '../src/core/atom'
+
+it('atom() should return the correct types', () => {
+  function Component() {
+    // primitive atom
+    const primitiveAtom = atom(0)
+    expectType<Atom<number> & WithInitialValue<number>>(primitiveAtom)
```
Comment (dai-shi) [inline tests/types.test.tsx:9]: Either,
```suggestion
    expectType<Atom<number> & { init: number }>(primitiveAtom)
```

Or,
```suggestion
    expectType<Atom<number>>(primitiveAtom)
```

What do you think?

Comment (dai-shi) [reply tests/types.test.tsx:9]: Also, we want to expect with `PrimitiveAtom` which is public type.

File: src/core/atom.ts (hunk 1 above)
Comment (Thisen) [reply src/core/atom.ts:20]: Then I can't test the type of `atom()`

Comment (dai-shi) [reply src/core/atom.ts:20]: I think what we want to test is `atom(0)` has `PrimitiveAtom<number>` type. That's it. WithInitialValue is implementation details and should be tested as behavior.

**Comment by Thisen** [comment]
@dai-shi Ready for review. :-)

**Review by dai-shi — APPROVED** [review]
Cool. This will be a good addition to make it maintainable.
We could add more for async atoms.
Great work. Thanks.

**Pull request merged**